### PR TITLE
Support sparse matrices in build_matrix

### PR DESCRIPTION
**Pull request by nilsvu**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/LinearSolver/BuildMatrix.hpp [hunk 1]
```diff
@@ -39,13 +40,17 @@ CREATE_IS_CALLABLE_V(reset)
...
 template <typename LinearOperator, typename OperandType, typename ResultType,
-          typename... OperatorArgs>
-void build_matrix(
-    const gsl::not_null<DenseMatrix<double, blaze::columnMajor>*> matrix,
-    const gsl::not_null<OperandType*> operand_buffer,
-    const gsl::not_null<ResultType*> result_buffer,
-    const LinearOperator& linear_operator,
-    const std::tuple<OperatorArgs...>& operator_args = {}) {
+          typename MT, bool SO, typename... OperatorArgs>
+void build_matrix(blaze::Matrix<MT, SO>& matrix,
```
Comment (nilsvu) [inline src/NumericalAlgorithms/LinearSolver/BuildMatrix.hpp]: @wthrowe I opened this draft PR to ask for your input: Is there a way to keep the `gsl::not_null` here? `blaze::Matrix` is the superclass of Blaze's dense and sparse matrix classes, and this `build_matrix` function should work with any of them. Blaze implements functions like this using non-const references (see e.g. their [solve](https://bitbucket.org/blaze-lib/blaze/wiki/Matrix%20Operations#!linear-systems) function), but I would prefer to make the non-const return argument explicit using `gsl::not_null`. However, the compiler doesn't resolve subclasses when I just write `const gsl::not_null<blaze::Matrix<MT, SO>*> matrix` for this argument. Can you think of a solution here?

Comment (wthrowe) [reply src/NumericalAlgorithms/LinearSolver/BuildMatrix.hpp]: I don't see anything clean.  It's easy, of course, if you are willing to name the base class at the call site, but that's pretty annoying.

If there's a generic way to get `MT` and `SO` from the derived classes, you could take a general `gsl::not_null<T*>` and do the conversion internally.  (Possibly in a simple wrapper function.)  Or, if `MT` and `SO` have very few possible values you could write overloads for each of them that convert and forward to the real implementation.

Comment (nilsvu) [reply src/NumericalAlgorithms/LinearSolver/BuildMatrix.hpp]: Thanks, I think your suggestion to take `gsl::not_null<T*>` is quite clean and works well 👍

**Comment by nilsvu** [comment]
This is ready for review now, and quite small. @wthrowe and @knelli2 would you do me the honor?

**Review by wthrowe — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]
LGTM

**Pull request merged**